File: sdk/rust/src/reasoning.rs
use crate::AlpError;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReasoningStep {
    pub step_id: String,
    pub agent_id: String,
    pub thought: String,
    pub action: String,
    pub observation: Option<String>,
    pub confidence: f64,
    pub dependencies: Vec<String>,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReasoningChain {
    pub chain_id: String,
    pub goal: String,
    pub steps: Vec<ReasoningStep>,
    pub created_at: String,
    pub status: String,
    pub result: Option<String>,
}

pub struct ReasoningTracer {
    chains: HashMap<String, ReasoningChain>,
    step_counter: usize,
}

impl ReasoningTracer {
    pub fn new() -> Self {
        Self {
            chains: HashMap::new(),
            step_counter: 0,
        }
    }

    pub fn create_chain(&mut self, goal: &str) -> ReasoningChain {
        let chain_id = format!(
            "chain-{}-{}",
            chrono::Utc::now().timestamp_millis(),
            rand_suffix(5)
        );
        let chain = ReasoningChain {
            chain_id: chain_id.clone(),
            goal: goal.to_string(),
            steps: Vec::new(),
            created_at: chrono::Utc::now().to_rfc3339(),
            status: "draft".to_string(),
            result: None,
        };
        self.chains.insert(chain_id.clone(), chain.clone());
        chain
    }

    pub fn add_step(
        &mut self,
        chain_id: &str,
        input: &ReasoningStep,
    ) -> Result<ReasoningStep, AlpError> {
        let chain = self
            .chains
            .get_mut(chain_id)
            .ok_or_else(|| AlpError::new(format!("Reasoning chain '{}' not found", chain_id)))?;
        if chain.status != "executing" {
            chain.status = "executing".to_string();
        }
        self.step_counter += 1;
        let step = ReasoningStep {
            step_id: format!("step-{}-{}", chain_id, self.step_counter),
            agent_id: input.agent_id.clone(),
            thought: input.thought.clone(),
            action: input.action.clone(),
            observation: input.observation.clone(),
            confidence: input.confidence,
            dependencies: input.dependencies.clone(),
            timestamp: chrono::Utc::now().to_rfc3339(),
        };
        chain.steps.push(step.clone());
        Ok(step)
    }

    pub fn complete_chain(
        &mut self,
        chain_id: &str,
        result: &str,
    ) -> Result<ReasoningChain, AlpError> {
        let chain = self
            .chains
            .get_mut(chain_id)
            .ok_or_else(|| AlpError::new(format!("Reasoning chain '{}' not found", chain_id)))?;
        chain.status = "completed".to_string();
        chain.result = Some(result.to_string());
        Ok(chain.clone())
    }

    pub fn fail_chain(&mut self, chain_id: &str, reason: &str) -> Result<ReasoningChain, AlpError> {
        let chain = self
            .chains
            .get_mut(chain_id)
            .ok_or_else(|| AlpError::new(format!("Reasoning chain '{}' not found", chain_id)))?;
        chain.status = "failed".to_string();
        chain.result = Some(reason.to_string());
        Ok(chain.clone())
    }

    pub fn get_chain(&self, chain_id: &str) -> Option<&ReasoningChain> {
        self.chains.get(chain_id)
    }

    pub fn get_steps_by_agent(&self, agent_id: &str) -> Vec<&ReasoningStep> {
        let mut steps = Vec::new();
        for chain in self.chains.values() {
            for step in &chain.steps {
                if step.agent_id == agent_id {
                    steps.push(step);
                }
            }
        }
        steps
    }
}

fn rand_suffix(n: usize) -> String {
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
    let mut s = String::with_capacity(n);
    for _ in 0..n {
        let idx = (chrono::Utc::now().timestamp_millis() as usize) % CHARS.len();
        s.push(CHARS[idx] as char);
    }
    s
}

File: sdk/go/reasoning.go
package alpgo

import (
	"fmt"
	"strings"
	"time"
)

type ReasoningStep struct {
	StepID      string   `json:"step_id"`
	AgentID     string   `json:"agent_id"`
	Thought     string   `json:"thought"`
	Action      string   `json:"action"`
	Observation string   `json:"observation,omitempty"`
	Confidence  float64  `json:"confidence"`
	Dependencies []string `json:"dependencies"`
	Timestamp   string   `json:"timestamp"`
}

type ReasoningChain struct {
	ChainID   string           `json:"chain_id"`
	Goal      string           `json:"goal"`
	Steps     []*ReasoningStep `json:"steps"`
	CreatedAt string           `json:"created_at"`
	Status    string           `json:"status"`
	Result    string           `json:"result,omitempty"`
}

type ReasoningTracer struct {
	chains       map[string]*ReasoningChain
	stepCounter  int
}

func NewReasoningTracer() *ReasoningTracer {
	return &ReasoningTracer{
		chains: make(map[string]*ReasoningChain),
	}
}

func (r *ReasoningTracer) CreateChain(goal string) *ReasoningChain {
	chainID := fmt.Sprintf("chain-%d-%s", time.Now().UnixNano(), randomSuffix(5))
	chain := &ReasoningChain{
		ChainID:   chainID,
		Goal:      goal,
		Steps:     []*ReasoningStep{},
		CreatedAt: time.Now().UTC().Format(time.RFC3339),
		Status:    "draft",
	}
	r.chains[chainID] = chain
	return chain
}

func (r *ReasoningTracer) AddStep(chainID string, input ReasoningStep) (*ReasoningStep, error) {
	chain, ok := r.chains[chainID]
	if !ok {
		return nil, fmt.Errorf("reasoning chain '%s' not found", chainID)
	}
	if chain.Status != "executing" {
		chain.Status = "executing"
	}
	r.stepCounter++
	step := &ReasoningStep{
		StepID:      fmt.Sprintf("step-%s-%d", chainID, r.stepCounter),
		AgentID:     input.AgentID,
		Thought:     input.Thought,
		Action:      input.Action,
		Observation: input.Observation,
		Confidence:  input.Confidence,
		Dependencies: input.Dependencies,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
	chain.Steps = append(chain.Steps, step)
	return step, nil
}

func (r *ReasoningTracer) CompleteChain(chainID, result string) (*ReasoningChain, error) {
	chain, ok := r.chains[chainID]
	if !ok {
		return nil, fmt.Errorf("reasoning chain '%s' not found", chainID)
	}
	chain.Status = "completed"
	chain.Result = result
	return chain, nil
}

func (r *ReasoningTracer) FailChain(chainID, reason string) (*ReasoningChain, error) {
	chain, ok := r.chains[chainID]
	if !ok {
		return nil, fmt.Errorf("reasoning chain '%s' not found", chainID)
	}
	chain.Status = "failed"
	chain.Result = reason
	return chain, nil
}

func (r *ReasoningTracer) GetChain(chainID string) (*ReasoningChain, bool) {
	chain, ok := r.chains[chainID]
	return chain, ok
}

func (r *ReasoningTracer) GetStepsByAgent(agentID string) []*ReasoningStep {
	var steps []*ReasoningStep
	for _, chain := range r.chains {
		for _, step := range chain.Steps {
			if step.AgentID == agentID {
				steps = append(steps, step)
			}
		}
	}
	return steps
}

func randomSuffix(n int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	var b strings.Builder
	for i := 0; i < n; i++ {
		b.WriteByte(chars[time.Now().UnixNano()%int64(len(chars))])
	}
	return b.String()
}
